lib/Storage/MongoDbTimeSeriesStore.php

```diff
@@ -120,137 +127,231 @@ public function setDependencies($log, $date, $layoutFactory = null, $campaignFac
+            }
+        }
+
+        // Widget name
+        $widgetName = null;
+        if (!empty($statData['widgetId'])) {
+
+            if (array_key_exists($statData['widgetId'], $this->widgetIds)) {
+
+                $widgetName = $this->widgetIds[$statData['widgetId']]['name'];
+
+            } else {
+
                 try {
-                    $widget = $this->widgetFactory->getById($stat['widgetId']);
+                    $widget = $this->widgetFactory->getById($statData['widgetId']);
+
+                    if($widget != null) {
+                        $widget->load();
+                        $widgetName = isset($mediaName) ? $mediaName : $widget->getOptionValue('name', $widget->type);
+
+                        // Put widget name to memory
+                        $this->widgetIds[$statData['widgetId']]['name'] = $widgetName;
+                    }
+
                 } catch (NotFoundException $error) {
                     // Widget not found, ignore and log the stat
-                    $this->log->error('Widget not found. Widget Id: '. $stat['widgetId'] .',Layout Id: '. $stat['layoutId']
-                        .', FromDT: '.$statData[$k]['start'].', ToDt: '.$statData[$k]['end'].', Type: '.$stat['type']
-                        .', Duration: '.$stat['duration'] .', Count '.$stat['count']);
+                    $this->log->error('Widget not found. Widget Id: '. $statData['widgetId'] .',Layout Id: '. $statData['layoutId']
+                        .', FromDT: '.$statData['start'].', ToDt: '.$statData['end'].', Type: '.$statData['type']
+                        .', Duration: '.$statData['duration'] .', Count '.$statData['count']);
 
-                    // unset($statData[$k]);
-                    continue;
-                }
-
-                if($widget != null) {
-                    $widget->load();
-                    $widgetName = isset($mediaName) ? $mediaName : $widget->getOptionValue('name', $widget->type);
-                    $statData[$k]['widgetName'] = $widgetName;
+                    return;
                 }
             }
 
-            // Display name
-            try {
-                $display = $this->displayFactory->getById($stat['displayId']); //TODO what if you dont find the displayId
-            } catch (NotFoundException $error) {
-                // Display not found, ignore and log the stat
-                $this->log->error('Display not found. Display Id: '. $stat['displayId'] .',Layout Id: '. $stat['layoutId']
-                    .', FromDT: '.$statData[$k]['start'].', ToDt: '.$statData[$k]['end'].', Type: '.$stat['type']
-                    .', Duration: '.$stat['duration'] .', Count '.$stat['count']);
-
-                // unset($statData[$k]);
-                continue;
-            }
+            // SET widgetName
+            $statData['widgetName'] = $widgetName;
 
-            $statData[$k]['displayName'] = $display->display;
+        }
+
+        // Display name
+        try {
```

Comment:
I think we pass this in from the Soap.php using `$this->display->displayId` - which we could change to `$this-display` and completely remove this lookup?